walkdir.py

```diff
@@ -276,10 +287,21 @@ def file_paths(walk_iter):
 
 def all_paths(walk_iter):
     """Iterate over both files and directories visited by the underlying walk"""
-    for dirpath, subdirs, files in walk_iter:
-        yield dirpath
-        for fname in files:
+    dir_entry = next(walk_iter, None)
+    if dir_entry is None:
+        return
+    top = dir_entry[0]
+    yield top
+    while dir_entry:
+        dirpath = dir_entry[0]
+        if top not in dirpath:
```

Comment:
This is a good catch, as the current tests don't cover underlying walks with multiple independent roots - it would be good to have a test case that strings together fake walks for "root/subdir1", "root/subdir2" and "root/other" with itertools.chain and checks the concatenated output is the same as that for walking "root", only with the root directory itself missing.

For the actual check itself, might `if not dirpath.startswith(top):` be a better way of checking if the underlying walk has switched to a different base path?